File: 01-Programming-Languages/C++/Data-Structures/Tree/binary_tree_traversal.cpp
/*
Problem: Binary Tree Traversals (Inorder, Preorder, Postorder)
Perform inorder, preorder, and postorder traversal of a binary tree.

Approach:
- Recursive depth-first traversal

Time Complexity: O(n)
Space Complexity: O(n)

Example:
Input: Tree: 1 -> (2, 3)
Output: Inorder: 2 1 3, Preorder: 1 2 3, Postorder: 2 3 1
*/

#include <iostream>
using namespace std;

struct Node
{
    int data;
    Node *left, *right;
    Node(int val) : data(val), left(nullptr), right(nullptr) {}
};

void inorder(Node *r)
{
    if (!r) return;
    inorder(r->left);
    cout << r->data << " ";
    inorder(r->right);
}

void preorder(Node *r)
{
    if (!r) return;
    cout << r->data << " ";
    preorder(r->left);
    preorder(r->right);
}

void postorder(Node *r)
{
    if (!r) return;
    postorder(r->left);
    postorder(r->right);
    cout << r->data << " ";
}

int main()
{
    Node *root = new Node(1);
    root->left = new Node(2);
    root->right = new Node(3);
    cout << "Inorder: "; inorder(root); cout << endl;
    cout << "Preorder: "; preorder(root); cout << endl;
    cout << "Postorder: "; postorder(root); cout << endl;
    return 0;
}

File: 01-Programming-Languages/PHP/Data-Structures/Tree/binary_tree_traversal.php
<?php

/*
Problem: Binary Tree Traversals
Description: Perform inorder, preorder, and postorder traversals on a binary tree.

Approach:
- Inorder: left -> root -> right
- Preorder: root -> left -> right
- Postorder: left -> right -> root

Time Complexity: O(n)
Space Complexity: O(n) (recursion stack)

Example:
Input:  Tree: 1 -> (2, 3) where 2 -> (4, 5)
Output:
  Inorder:   4 2 5 1 3
  Preorder:  1 2 4 5 3
  Postorder: 4 5 2 3 1
*/

class TreeNode {
    public $val;
    public $left;
    public $right;

    public function __construct($val = 0, $left = null, $right = null) {
        $this->val = $val;
        $this->left = $left;
        $this->right = $right;
    }
}

function inorder(?TreeNode $root, array &$result): void {
    if ($root === null) return;
    inorder($root->left, $result);
    $result[] = $root->val;
    inorder($root->right, $result);
}

function preorder(?TreeNode $root, array &$result): void {
    if ($root === null) return;
    $result[] = $root->val;
    preorder($root->left, $result);
    preorder($root->right, $result);
}

function postorder(?TreeNode $root, array &$result): void {
    if ($root === null) return;
    postorder($root->left, $result);
    postorder($root->right, $result);
    $result[] = $root->val;
}

$root = new TreeNode(1,
    new TreeNode(2, new TreeNode(4), new TreeNode(5)),
    new TreeNode(3)
);

$in = []; inorder($root, $in);
$pre = []; preorder($root, $pre);
$post = []; postorder($root, $post);

echo "Inorder:   " . implode(" ", $in) . "\n";
echo "Preorder:  " . implode(" ", $pre) . "\n";
echo "Postorder: " . implode(" ", $post) . "\n";
